flink-runtime/src/main/java/org/apache/flink/runtime/state/StateUtil.java

```diff
@@ -63,10 +65,16 @@ public static void bestEffortDiscardAllStateObjects(
 	public static void discardStateFuture(RunnableFuture<? extends StateObject> stateFuture) throws Exception {
 		if (null != stateFuture) {
 			if (!stateFuture.cancel(true)) {
-				StateObject stateObject = FutureUtil.runIfNotDoneAndGet(stateFuture);
 
-				if (null != stateObject) {
-					stateObject.discardState();
+				try {
+					// We attempt to get a result, in case the future completed before cancellation.
+					StateObject stateObject = FutureUtil.runIfNotDoneAndGet(stateFuture);
+
+					if (null != stateObject) {
+						stateObject.discardState();
+					}
+				} catch (CancellationException | ExecutionException ignore) {
+					// No result that requires discarding was produced.
```

Comment:
Maybe we should log on DEBUG that the state future could not be completed.

Reply:
Ok, but I wonder if we should include the exception to the log entry, because I want to avoid presenting something to the user that might look like a problem/bug.